src/cljs/owlet_ui/components/search_bar.cljs

```diff
@@ -8,41 +8,72 @@
 
 (defonce suggestion-count (reagent/atom 16))
 
+(defonce this-scroll (atom 0))
+
+(defonce last-scroll (atom 0))
+
 (defn toggle-suggestions []
   (if-let [suggestions (aget (js->clj (js/document.getElementsByClassName "rc-typeahead-suggestions-container")) 0)]
     (let [hidden (.-hidden suggestions)]
       (when-not (nil? suggestions)
         (set! (.-hidden suggestions) (not hidden))))))
 
+(defn show-search []
+  (let [search (aget (js->clj (js/document.getElementsByClassName "form-control")) 0)]
+    (set! (-> search .-style .-height) "50px")
```

Comment:
Yes, @ZadenRB, at some point, we'll need to rewrite search_bar.cljs in a functional way. Simply put, we want to minimize the use of mutation, so the fewer atoms and exclamation points you see, the better.

The JavaScript/JQuery way is to assemble a bunch of DOM elements and variables, then mutate them when things happen. Your code is attempting to do the same thing, except with atoms. The re-frame way is to define functions that describe the tree of components, which constantly get called by re-frame when things happen. Then re-frame handles the mutation of DOM elements as needed.

Ideally, no atoms should be necessary, but sometimes local state is appropriate. Even then, however, we can minimize them by thinking deeply about what state is really needed.

Here's a simple example: You're using two atoms, `this-scroll` and `last-scroll`, when just one will do, which I call `scroll-delta` below.

    (def scroll-delta (atom `(0 0)))
    
    ; push-scroll is a function: (push-scroll '(22 11) 33) gives '(33 22) .
    ;
    (def push-scroll (comp (partial drop-last 1) conj))
    
    (defn change-scroll!
      [n]
      (swap! scroll-delta push-scroll n))
      ; Here, swap! calculated the new value as (push-scroll <old-value> n)

So if you did this:

    (do (change-scroll! 20) (change-scroll! 30) (change-scroll! 40))

then `@scroll-delta` will end up as `'(40 30)`. Plus, your calculations will be neater:

    (when (>= (- @this-scroll @last-scroll) 50)
      (hide-search))
    (when (<= (- @this-scroll @last-scroll) -50)
      (show-search)))

becomes

    (let [delta (apply - @scroll-delta)]
      (cond
        (<= 50 delta) (hide-search)
        (<= delta -50) (show-search)))

which perhaps belongs in a helpfully named function of its own.

This is a small example of how we can ensure that the "this-scroll/last-scroll" relationship is always correct, since they are defined in the same atom. This idea is just "encapsulation" in OOP.

Does this strike you as pedantic? It may seem slow and ponderous to have to think so deeply about what you hoped _should_ be quick and easy! But minimizing and carefully managing state not only helps keep the code robust, but it helps a future programmer (maybe you) understand how it works. **That's** really the most important goal.